store/sqlstore/channel_store.go

```diff
@@ -3950,46 +3948,175 @@ func (s SqlChannelStore) UpdateSidebarCategories(userId, teamId string, categori
-		}
+
 		// if new preference is false - remove the channel from the appropriate sidebar category
 		if preference.Value == "false" {
-			var deleteQuery string
-			if s.DriverName() == model.DATABASE_DRIVER_MYSQL {
-				deleteQuery = "DELETE SidebarChannels FROM SidebarChannels LEFT JOIN SidebarCategories ON SidebarCategories.Id = SidebarChannels.CategoryId WHERE SidebarCategories.Type=:CategoryType AND SidebarCategories.UserId=:UserId AND SidebarChannels.UserId=:UserId AND ChannelId=:ChannelId"
-			} else {
-				deleteQuery = "DELETE FROM SidebarChannels USING SidebarChannels AS chan LEFT OUTER JOIN SidebarCategories AS cat ON cat.Id = chan.CategoryId WHERE cat.Type=:CategoryType AND cat.UserId = :UserId AND chan.UserId = :UserId AND   cat.TeamId = :TeamId AND chan.ChannelId=:ChannelId"
-			}
-
-			if _, err := transaction.Exec(deleteQuery, params); err != nil {
-				return model.NewAppError("SqlChannelStore.UpdateSidebarChannelByPreference", "store.sql_channel.sidebar_categories.app_error", nil, err.Error(), http.StatusInternalServerError)
+			if err := s.removeSidebarEntriesForPreferenceT(transaction, &preference); err != nil {
+				return err
 			}
 		} else {
-			// otherwise - insert new channel into the apropriate category. ignore duplicate error
-			if _, err := transaction.Exec("INSERT INTO SidebarChannels (ChannelId, UserId, CategoryId, SortOrder) SELECT Id AS CategoryId, :UserId AS UserId, :ChannelId AS ChannelId, MAX(SidebarChannels.SortOrder)+10 FROM SidebarCategories INNER JOIN SidebarChannels ON SidebarChannels.CategoryId = SidebarCategories.Id WHERE SidebarCategories.Type=:CategoryType AND SidebarCategories.UserId=:UserId GROUP BY SidebarChannels.CategoryId, SidebarCategories.Id", params); err != nil && !IsUniqueConstraintError(err, []string{"UserId"}) {
-				return model.NewAppError("SqlChannelStore.UpdateSidebarChannelByPreference", "store.sql_channel.sidebar_categories.app_error", nil, err.Error(), http.StatusInternalServerError)
+			if err := s.addChannelToFavoritesCategory(transaction, &preference); err != nil {
+				return err
 			}
 		}
 	}
 
+	return transaction.Commit()
+}
+
+func (s SqlChannelStore) removeSidebarEntriesForPreferenceT(transaction *gorp.Transaction, preference *model.Preference) error {
+	if preference.Category != model.PREFERENCE_CATEGORY_FAVORITE_CHANNEL {
+		return nil
+	}
+
+	// Delete any corresponding SidebarChannels entries in a Favorites category corresponding to this preference. This
+	// can't use the query builder because it uses DB-specific syntax
+	params := map[string]interface{}{
+		"UserId":       preference.UserId,
+		"ChannelId":    preference.Name,
+		"CategoryType": model.SidebarCategoryFavorites,
+	}
+	var query string
+	if s.DriverName() == model.DATABASE_DRIVER_MYSQL {
+		query = `
+			DELETE
+				SidebarChannels
+			FROM
+				SidebarChannels
+			JOIN
+				SidebarCategories ON SidebarChannels.CategoryId = SidebarCategories.Id
+			WHERE
+				UserId = :UserId
```

Comment:
It wasn't when I was testing, but it's probably good to change that to avoid accidental ambiguity. The UserId existing on both tables is redundant, so I wouldn't worry about it being different though.

Reply:
Sounds good -- is the plan to deprecate one of the `UserId` columns, or do we use both and just rely on them never diverging?

Reply:
I think it'll probably be the latter. I had originally intended on having some uniqueness constraints on those keys, but with how entries are added/removed from the sidebar one at a time, that wasn't possible.